components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/UnifiedClaimMetadataManager.java

```diff
@@ -302,15 +302,25 @@ public List<ExternalClaim> getExternalClaims(String externalClaimDialectURI, int
 
         Map<String, ExternalClaim> externalClaimsInDBMap = externalClaimsInDB.stream()
                 .collect(Collectors.toMap(ExternalClaim::getClaimURI, claim -> claim));
-
+        Map<String, ExternalClaim> localClaimInDBMap = externalClaimsInDB.stream()
+                .collect(Collectors.toMap(ExternalClaim::getMappedLocalClaim, claim -> claim));
+
+        /*
+         * If a system claim is also in the DB, then the claim retrieved from the DB gets the priority.
+         * Also, if there is a system claim that is mapped to the same local claim as another external claim in the same
+         * dialect, then we do not enforce the system claim on the tenant because that would violate the constraint of
+         * having a unique claim mapping within the dialect. This is to preserve backward-compatibility.
+         */
         List<ExternalClaim> allExternalClaims = new ArrayList<>();
         for (ExternalClaim externalClaimInSystem : externalClaimsInSystem) {
             ExternalClaim matchingClaimInDB = externalClaimsInDBMap.get(externalClaimInSystem.getClaimURI());
+            ExternalClaim externalClaimWithSameLocalClaim = localClaimInDBMap.get(
+                    externalClaimInSystem.getMappedLocalClaim());
             if (matchingClaimInDB != null) {
-                matchingClaimInDB.setClaimProperty(ClaimConstants.IS_SYSTEM_CLAIM, Boolean.TRUE.toString());
+                markAsSystemClaim(matchingClaimInDB);
                 allExternalClaims.add(matchingClaimInDB);
                 externalClaimsInDBMap.remove(externalClaimInSystem.getClaimURI());
-            } else {
+            } else if (externalClaimWithSameLocalClaim == null) {
```

Comment:
Shall we use `!localClaimInDBMap.containsKey(externalClaimInSystem.getMappedLocalClaim())` here? So that we can avoid defining a new variable for this.

Reply:
Fixed with 7004a1f7b74d902e3a30e3aa7b86a30587c520b1